src/Documentation complete/man/man3/visa_C.3
.TH "visa_C" 3 "Mercredi Octobre 25 2017" "Simulateur Documentation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
visa_C \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <visa_C\&.h>\fP
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "void \fBOpen\fP (\fBvisaDefaultRM_C\fP *viRM, char instr_i[])"
.br
.ti -1c
.RI "void \fBReadBinary\fP (char *buf_i, long sizeBuf_i, unsigned long *retCnt_i)"
.br
.ti -1c
.RI "void \fBReadAscii\fP (string *data_i)"
.br
.ti -1c
.RI "void \fBReadAscii\fP (string *data_i, unsigned int bufferLength_i)"
.br
.ti -1c
.RI "void \fBWriteBinary\fP (const char *data_i, int size_i)"
.br
.ti -1c
.RI "void \fBWriteAscii\fP (const string *data_i)"
.br
.ti -1c
.RI "void \fBWriteAscii\fP (const char *data_i)"
.br
.ti -1c
.RI "void \fBClear\fP ()"
.br
.ti -1c
.RI "long \fBReadSTB\fP ()"
.br
.ti -1c
.RI "void \fBClose\fP ()"
.br
.ti -1c
.RI "void \fBSetAttribute\fP (long attr_i, long attrstate_i)"
.br
.ti -1c
.RI "void \fBGetAttribute\fP (long attr_i, void _VI_PTR attrstate_i)"
.br
.ti -1c
.RI "void \fBFlush\fP (long mask_i)"
.br
.ti -1c
.RI "void \fBSetBuf\fP (long mask_i, long size_i)"
.br
.ti -1c
.RI "void \fBWait\fP (int Delay_ms_i=100)"
.br
.ti -1c
.RI "void \fBOperationComplete\fP (int Timeout_ms_i=300000)"
.br
.ti -1c
.RI "void \fBsetDebug\fP (bool debug_i=true)"
.br
.ti -1c
.RI "virtual void \fBSetTimeOutMs\fP (long timeOutMs)"
.br
.ti -1c
.RI "\fB~visa_C\fP ()"
.br
.ti -1c
.RI "\fBvisa_C\fP ()"
.br
.ti -1c
.RI "ViStatus \fBviSetAttribute\fP (long attr_i, ViAttrState attrstate_i)"
.br
.ti -1c
.RI "ViStatus \fBviGetAttribute\fP (long attr_i, void _VI_PTR attrstate_i)"
.br
.ti -1c
.RI "ViStatus \fBviRead\fP (ViPBuf buf, ViUInt32 cnt, ViPUInt32 retCnt)"
.br
.ti -1c
.RI "ViStatus \fBviWrite\fP (ViBuf buf, ViUInt32 cnt, ViPUInt32 retCnt)"
.br
.ti -1c
.RI "ViStatus \fBviPrintf\fP (ViString writeFmt, void _VI_PTR param)"
.br
.ti -1c
.RI "ViStatus \fBviQueryf\fP (ViString writeFmt, ViString readFmt)"
.br
.ti -1c
.RI "ViStatus \fBviQueryf\fP (ViString writeFmt, ViString readFmt, void _VI_PTR param)"
.br
.ti -1c
.RI "ViStatus \fBviQueryf\fP (ViString writeFmt, ViString readFmt, void _VI_PTR param, void _VI_PTR param2)"
.br
.ti -1c
.RI "ViStatus \fBviScanf\fP (ViString readFmt, void _VI_PTR param)"
.br
.ti -1c
.RI "ViStatus \fBviStatusDesc\fP (ViStatus status, ViChar _VI_FAR desc[])"
.br
.ti -1c
.RI "ViStatus \fBviReadSTB\fP (ViPUInt16 status)"
.br
.ti -1c
.RI "ViStatus \fBviFlush\fP (ViUInt16 mask)"
.br
.in -1c
.SS "Attributs publics"

.in +1c
.ti -1c
.RI "unsigned long \fBsession_m\fP"
.br
.in -1c
.SH "Description détaillée"
.PP 
Classe Visa niveau instrument\&. 
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "\fBvisa_C::~visa_C\fP ()"
Destructeur, fermeture implicite de session de carte\&. 
.SS "\fBvisa_C::visa_C\fP ()"
Constructeur par défaut\&. 
.SH "Documentation des fonctions membres"
.PP 
.SS "void \fBvisa_C::Clear\fP ()"
Effacement\&. 
.SS "void \fBvisa_C::Close\fP ()"
Fermeture session instrument\&. 
.SS "void \fBvisa_C::Flush\fP (longmask_i)"
Gestion du mack\&. 
.PP
\fBParamètres:\fP
.RS 4
\fImask_i\fP paramètre d'entrée définissant le mask\&. 
.RE
.PP

.SS "void \fBvisa_C::GetAttribute\fP (longattr_i, void _VI_PTRattrstate_i)"
Lecture des attributs\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIattr_i\fP paramètre d'entrée définissant l'attribut à manipuler\&. 
.RE
.PP

.SS "void \fBvisa_C::Open\fP (\fBvisaDefaultRM_C\fP *viRM, charinstr_i[])"
Ouverture session instrument\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIviRM\fP pointeur sur la session VISA d'accés à la carte GPIB\&. 
.br
\fIinstr_i\fP pointeur sur le premier caractère de l'adresse de l'instrument\&. 
.RE
.PP

.SS "void \fBvisa_C::OperationComplete\fP (intTimeout_ms_i = \fC300000\fP)"
Attente fin d'opération instrument\&. 
.PP
\fBParamètres:\fP
.RS 4
\fITimeout_ms_i\fP paramètre d'entrée optionnel définissant la durée de l'attente en ms avant de remonter un timeout, par défaut cette durée de timeout est de 5min\&. 
.RE
.PP

.SS "void \fBvisa_C::ReadAscii\fP (string *data_i)"
Lecture en ASCII sous la forme string\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIdata_i\fP pointeur sur la chaine de string retournée par la fonction\&. 
.RE
.PP

.SS "void \fBvisa_C::ReadAscii\fP (string *data_i, unsigned intbufferLength_i)"
Lecture en ASCII sous la forme char\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIdata_i\fP pointeur sur le premier caractère de la chaine de string retournée par la fonction\&. 
.br
\fIbufferLength_i\fP paramètre d'entrée définissant la longeur du buffer alloué en entrée de la fonction\&. 
.RE
.PP

.SS "void \fBvisa_C::ReadBinary\fP (char *buf_i, longsizeBuf_i, unsigned long *retCnt_i)"
Lecture en binaire\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIbuf_i\fP pointeur sur le buffer retourné par la fonction\&. 
.br
\fIsizeBuf_i\fP paramètre d'entrée définissant la longeur du buffer alloué en entrée de la fonction\&. 
.br
\fIretCnt_i\fP pointeur sur le nombre de caractères retournés par la fonction\&. 
.RE
.PP

.SS "long \fBvisa_C::ReadSTB\fP ()"
Lecture de l'octet de status, retourne un long\&. 
.SS "void \fBvisa_C::SetAttribute\fP (longattr_i, longattrstate_i)"
Ecriture des attributs\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIattr_i\fP paramètre d'entrée définissant l'attribut à manipuler\&. 
.br
\fIattrstate_i\fP paramètre d'entrée définissant l'état à donner à l'attribut\&. 
.RE
.PP

.SS "void \fBvisa_C::SetBuf\fP (longmask_i, longsize_i)"
Ecriture du buffer\&. 
.PP
\fBParamètres:\fP
.RS 4
\fImask_i\fP paramètre d'entrée définissant le mask\&. 
.RE
.PP

.SS "void \fBvisa_C::setDebug\fP (booldebug_i = \fCtrue\fP)"
Activation / désaction debug\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIdebug_i\fP paramètre d'entrée optionnel définissant si on est en mode debug ou pas, par défaut la couche VISA fonctionne en mode debug\&. 
.RE
.PP

.SS "void \fBvisa_C::SetTimeOutMs\fP (longtimeOutMs)\fC [virtual]\fP"
Positionnement du time-out de reception des donnees 
.PP
\fBParamètres:\fP
.RS 4
\fItimeOutMs\fP time-out en milliseconde 
.RE
.PP

.SS "void \fBvisa_C::Wait\fP (intDelay_ms_i = \fC100\fP)"
Attente\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIDelay_ms_i\fP paramètre d'entrée optionnel définissant la durée de l'attente en ms, par défaut cette durée est de 100ms\&. 
.RE
.PP

.SS "void \fBvisa_C::WriteAscii\fP (const string *data_i)"
Ecriture en ASCII sous la forme string\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIdata_i\fP paramètre d'entrée définissant la chaine de string en entrée de la fonction\&. 
.RE
.PP

.SS "void \fBvisa_C::WriteAscii\fP (const char *data_i)"
Ecriture en ASCII sous la forme char\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIdata_i\fP pointeur sur le premier caractère de la chaine de char en entrée de la fonction\&. 
.RE
.PP

.SS "void \fBvisa_C::WriteBinary\fP (const char *data_i, intsize_i)"
Ecriture en binaire\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIdata_i\fP pointeur sur le premier caractère à envoyer par la fonction\&. 
.br
\fIsize_i\fP paramètre d'entrée définissant la longeur de la chaine en entrée de la fonction\&. 
.RE
.PP

.SH "Documentation des données membres"
.PP 
.SS "unsigned long \fBvisa_C::session_m\fP"
Variable membre privée contenant le numéro de session VISA au niveau instrument\&. 

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour Simulateur Documentation à partir du code source\&.


src/Librairie/visa_C.cpp
// ---------------------------------------------------------------------------
#include <iostream>
#include "visa_C.h"
#include "errorInfo.h"
#ifndef linux
#pragma package(smart_init)
#endif

// ---------------------------------------------------------------------------

/* Conversion error number en chaine de caracteres */
static void errVisaToStr(long errVisa_i, char *msg_o, int maxMsg_i) {
  ERRORINFO_TRY_M

  char desc_l[BUFFER_LENGTH];

  if (viStatusDesc_s != NULL) {
    viStatusDesc_s(0, errVisa_i, desc_l);
    strncpy(msg_o, desc_l, maxMsg_i);
    msg_o[maxMsg_i - 1] = '\0';
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* --------------------Fonctions publiques----------------------- */

/* Ouverture de session visa niveau carte */
void visaDefaultRM_C::OpenDefaultRm(void) {
  ERRORINFO_TRY_M

  LibId = LoadLibrary("visa32.dll");
  if (LibId <= 0)
    throw Exception("La librairie visa32.dll est introuvable !");
  viOpenDefaultRM_s = (PTR_FUNCTION_viOpenDefaultRM)GetProcAddress(LibId, "viOpenDefaultRM");
  viOpen_s = (PTR_FUNCTION_viOpen)GetProcAddress(LibId, "viOpen");
  viClose_s = (PTR_FUNCTION_viClose)GetProcAddress(LibId, "viClose");
  viSetAttribute_s = (PTR_FUNCTION_viSetAttribute)GetProcAddress(LibId, "viSetAttribute");
  viGetAttribute_s = (PTR_FUNCTION_viGetAttribute)GetProcAddress(LibId, "viGetAttribute");
  viFlush_s = (PTR_FUNCTION_viFlush)GetProcAddress(LibId, "viFlush");
  viStatusDesc_s = (PTR_FUNCTION_viStatusDesc)GetProcAddress(LibId, "viStatusDesc");
  viRead_s = (PTR_FUNCTION_viRead)GetProcAddress(LibId, "viRead");
  viWrite_s = (PTR_FUNCTION_viWrite)GetProcAddress(LibId, "viWrite");
  viPrintf_s = (PTR_FUNCTION_viPrintf)GetProcAddress(LibId, "viPrintf");
  viQueryf_s = (PTR_FUNCTION_viQueryf)GetProcAddress(LibId, "viQueryf");
  viScanf_s = (PTR_FUNCTION_viScanf)GetProcAddress(LibId, "viScanf");
  viReadSTB_s = (PTR_FUNCTION_viReadSTB)GetProcAddress(LibId, "viReadSTB");
  viClear_s = (PTR_FUNCTION_viClear)GetProcAddress(LibId, "viClear");
  viEnableEvent_s = (PTR_FUNCTION_viEnableEvent)GetProcAddress(LibId, "viEnableEvent");
  viDisableEvent_s = (PTR_FUNCTION_viDisableEvent)GetProcAddress(LibId, "viDisableEvent");
  viDiscardEvents_s = (PTR_FUNCTION_viDiscardEvents)GetProcAddress(LibId, "viDiscardEvents");
  viWaitOnEvent_s = (PTR_FUNCTION_viWaitOnEvent)GetProcAddress(LibId, "viWaitOnEvent");
  viSetBuf_s = (PTR_FUNCTION_viSetBuf)GetProcAddress(LibId, "viSetBuf");

  errorInfoAddClass(VISA_ERR_VISA_CLASS, errVisaToStr);
  viOpenDefaultRM_s(&sess_m);

  ERRORINFO_CATCH_RETHROW_M
}

/* Destructeur de session visa niveau carte */
void visaDefaultRM_C::CloseDefaultRm(void) {
  ERRORINFO_TRY_M

  viClose_s(sess_m);
  FreeLibrary(LibId);

  ERRORINFO_CATCH_RETHROW_M
}

/* Constructeur de session au niveau instrumant */
visa_C::visa_C() {

  debug_m = false;
  deviceClosed_m = true;
}

/* Destructeur de session au niveau instrumant */
visa_C::~visa_C() {
  ERRORINFO_TRY_M

  if (!deviceClosed_m)
    viClose_s(session_m);

  deviceClosed_m = true;

  ERRORINFO_CATCH_RETHROW_M
}

/* Ouverture de session au niveau instrument */
void visa_C::Open(visaDefaultRM_C *viRM, char *instr_i) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  ret_l = viOpen_s(viRM->sess_m, instr_i, 0, 0, &session_m);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viOpen", "Unable to open visa ressource");
  }
  visaRessource_m = instr_i;
  deviceClosed_m = false;

  ERRORINFO_CATCH_RETHROW_M
}

/* Lecture en binaire */
void visa_C::ReadBinary(char *buf_i, long sizeBuf_i, unsigned long *retCnt_i) {
  ERRORINFO_TRY_M

  int ret_l;

  ret_l = viRead_s(session_m, (ViByte*)buf_i, sizeBuf_i, retCnt_i);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viRead", "Unable to read on visa ressource");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Lecture en ASCII */
void visa_C::ReadAscii(string *data_i) {
  ERRORINFO_TRY_M

  char buf_l[BUFFER_LENGTH];

  unsigned long retCnt_l = 0;

  this->ReadBinary(buf_l, sizeof(buf_l) / sizeof(buf_l[0]), &retCnt_l);
  buf_l[retCnt_l] = '\0';
  *data_i = (string)buf_l;
  if (debug_m) {
    cout << visaRessource_m << "<-" << *data_i << endl;
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Lecture en ASCII avec taille de buffer allouee dynamiquement */
void visa_C::ReadAscii(string *data_i, unsigned int bufferLength_i) {
  ERRORINFO_TRY_M

  char *buf_l;

  buf_l = new char[bufferLength_i];

  unsigned long retCnt_l = 0;

  this->ReadBinary(buf_l, bufferLength_i, &retCnt_l);
  buf_l[retCnt_l] = '\0';
  *data_i = buf_l;
  delete[]buf_l;

  ERRORINFO_CATCH_RETHROW_M
}

/* Ecriture en binaire */
void visa_C::WriteBinary(const char *data_i, int size_i) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  unsigned long retcnt_l = 0;
  char error_l[1024];

  ret_l = viWrite_s(session_m, (unsigned char *)data_i, size_i, &retcnt_l);
  if (ret_l < 0) {
    sprintf(error_l, "viWrite(%s)", data_i);
    throwError(VISA_ERR_VISA_CLASS, ret_l, error_l, "Unable to write to visa ressource");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Ecriture en ASCII sous forme string */
void visa_C::WriteAscii(const string *data_i) {
  ERRORINFO_TRY_M

  this->WriteBinary(data_i->c_str(), data_i->size());

  if (debug_m) {
    cout << visaRessource_m << "->" << *data_i << endl;
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Ecriture en ASCII suivant parametres sous forme char* */
void visa_C::WriteAscii(const char *data_i) {
  ERRORINFO_TRY_M

  this->WriteBinary(data_i, strlen(data_i));

  if (debug_m) {
    cout << visaRessource_m << "->" << data_i << endl;
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Effacement */
void visa_C::Clear() {
  ERRORINFO_TRY_M

  long ret_l = 0;

  ret_l = viClear_s(session_m);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viClear", "Unable to clear visa ressource");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Lecture STB */
long visa_C::ReadSTB() {
  ERRORINFO_TRY_M

  long ret_l = 0;

  unsigned short stat_l = 0;

  ret_l = viReadSTB_s(session_m, &stat_l);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viReadSTB", "Unable to read STB on visa ressource");
  }

  return stat_l;

  ERRORINFO_CATCH_RETHROW_M
}

/* Fermeture session instrument */
void visa_C::Close() {
  ERRORINFO_TRY_M

  long ret_l = 0;

  if (!deviceClosed_m)
    ret_l = viClose_s(session_m);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viClose", "Unable to close visa ressource");
  }

  deviceClosed_m = true;
  ERRORINFO_CATCH_RETHROW_M
}

/* Ecriture des attributs */
void visa_C::SetAttribute(long attr, long attrstate) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  ret_l = viSetAttribute_s(session_m, attr, attrstate);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viSetAttribute", "Unable to set visa ressource attribute");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Lecture des attributs */
void visa_C::GetAttribute(long attr, void _VI_PTR attrstate) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  ret_l = viGetAttribute_s(session_m, attr, attrstate);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viGetAttribute", "Unable to get visa ressource attribute");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Gestion du mask */
void visa_C::Flush(long mask_i) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  ret_l = viFlush_s(session_m, mask_i);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viFlush", "Unable to set visa mask");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Gestion du mask */
void visa_C::SetBuf(long mask_i, long size_i) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  ret_l = viSetBuf_s(session_m, mask_i, size_i);

  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viSetBuf", "Unable to set visa buffer");
  }

  ERRORINFO_CATCH_RETHROW_M
}

/* Attente */
void visa_C::Wait(int Delay_ms_i) {
  ERRORINFO_TRY_M

  viEnableEvent_s(session_m, VI_EVENT_SERVICE_REQ, VI_QUEUE, VI_NULL);
  viWaitOnEvent_s(session_m, VI_EVENT_SERVICE_REQ, Delay_ms_i, VI_NULL, VI_NULL);
  viDisableEvent_s(session_m, VI_EVENT_SERVICE_REQ, VI_QUEUE);
  // ReadSTB();

  ERRORINFO_CATCH_RETHROW_M
}

/* Attente OPC */
void visa_C::OperationComplete(int Timeout_ms_i) {
  ERRORINFO_TRY_M

  long ret_l = 0;

  viEnableEvent_s(session_m, VI_EVENT_SERVICE_REQ, VI_QUEUE, VI_NULL);
  ret_l = viWaitOnEvent_s(session_m, VI_EVENT_SERVICE_REQ, Timeout_ms_i, VI_NULL, VI_NULL);
  viDisableEvent_s(session_m, VI_EVENT_SERVICE_REQ, VI_QUEUE);
  ReadSTB();
  if (ret_l < 0) {
    throwError(VISA_ERR_VISA_CLASS, ret_l, "viOperationComplete", "Time-out occured before visa ressource complete operation");
  }

  ERRORINFO_CATCH_RETHROW_M
}

void visa_C::setDebug(bool debug_i) {
  debug_m = debug_i;
}

void visa_C::SetTimeOutMs(long timeOutMs) {
  ERRORINFO_TRY_M

  this->SetAttribute(VI_ATTR_TMO_VALUE, timeOutMs);

  ERRORINFO_CATCH_RETHROW_M
}

ViStatus visa_C::viSetAttribute(long attr_i, ViAttrState attrstate_i){
  return viSetAttribute_s(session_m, attr_i, attrstate_i);
}

ViStatus visa_C::viGetAttribute(long attr_i, void _VI_PTR attrstate_i){
  return viGetAttribute_s(session_m, attr_i, attrstate_i);
}

ViStatus visa_C::viRead(ViPBuf buf, ViUInt32 cnt, ViPUInt32 retCnt){
  return viRead_s(session_m, buf, cnt, retCnt);
}

ViStatus visa_C::viWrite(ViBuf  buf, ViUInt32 cnt, ViPUInt32 retCnt){
  return viWrite_s(session_m, buf, cnt, retCnt);
}

ViStatus visa_C::viPrintf(ViString writeFmt, void _VI_PTR param){
  return viPrintf_s(session_m, writeFmt, param);
}

ViStatus visa_C::viQueryf(ViString writeFmt, ViString readFmt){
  return viQueryf_s(session_m, writeFmt, readFmt);
}

ViStatus visa_C::viQueryf(ViString writeFmt, ViString readFmt, void _VI_PTR param){
  return viQueryf_s(session_m, writeFmt, readFmt, param);
}

ViStatus visa_C::viQueryf(ViString writeFmt, ViString readFmt, void _VI_PTR param, void _VI_PTR param2){
  return viQueryf_s(session_m, writeFmt, readFmt, param, param2);
}

ViStatus visa_C::viScanf(ViString readFmt, void _VI_PTR param){
  return viScanf_s(session_m, readFmt, param);
}

ViStatus visa_C::viStatusDesc(ViStatus status, ViChar _VI_FAR desc[]){
  return viStatusDesc_s(session_m, status, desc);
}

ViStatus visa_C::viReadSTB(ViPUInt16 status){
  return viReadSTB_s(session_m, status);
}

ViStatus visa_C::viFlush(ViUInt16 mask){
  return viFlush_s(session_m, mask);
}

